First step into a PIP world

shaftoe opened the pull request:
* To be able to reuse this code (especially the Nuimo class), we need something like this to let it be installed via PIP. This already works with

  `pip install https://github.com/shaftoe/nuimo-raspberrypi-demo/archive/pip.zip#nuimocore==0.0.1`

  and also avoids name collisions with other `nuimo` projects for example.

* update README to use PIP for bluepy install

* fix old_style class declaration

* add setup.py and .gitignore

larsblumberg commented:
Wow! Will try it out on my Raspberry and then merge it. Thanks a lot!

larsblumberg commented:
@shaftoe Thank you a lot for your efforts! I have a few questions to understand things before merging the PR:

- Can we name the library `nuimo-bluepy` instead of `nuimocore`? We are working on dedicated Linux/Python SDK that will make use of `pygattlib` instead of `bluepy`. That way we could have a distinct naming differentiation. What do you think?
- Can we have the `__init__.py` script in the root folder or is the subfolder necessary?
- As I am new to the Python universe, can you please explain why the `setup.py` script is needed?

larsblumberg commented:
Can you please add the build step

`sudo pip install -U pip`

before doing `sudo pip install https://github.com/IanHarvey/bluepy/archive/master.zip#bluepy`? The latter otherwise crashes on a freshly installed Raspbian with

```
pi@raspberrypi:~ $ sudo pip install https://github.com/IanHarvey/bluepy/archive/master.zip#bluepy
Downloading/unpacking https://github.com/IanHarvey/bluepy/archive/master.zip
  Downloading master.zip (unknown size): 462kB downloaded
Cleaning up...
Exception:
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/pip/basecommand.py", line 122, in main
    status = self.run(options, args)
  File "/usr/lib/python2.7/dist-packages/pip/commands/install.py", line 290, in run
    requirement_set.prepare_files(finder, force_root_egg_info=self.bundle, bundle=self.bundle)
  File "/usr/lib/python2.7/dist-packages/pip/req.py", line 1198, in prepare_files
    do_download,
  File "/usr/lib/python2.7/dist-packages/pip/req.py", line 1376, in unpack_url
    self.session,
  File "/usr/lib/python2.7/dist-packages/pip/download.py", line 572, in unpack_http_url
    download_hash = _download_url(resp, link, temp_location)
  File "/usr/lib/python2.7/dist-packages/pip/download.py", line 433, in _download_url
    for chunk in resp_read(4096):
  File "/usr/lib/python2.7/dist-packages/pip/download.py", line 421, in resp_read
    chunk_size, decode_content=False):
  File "/usr/lib/python2.7/dist-packages/urllib3/response.py", line 256, in stream
    data = self.read(amt=amt, decode_content=decode_content)
  File "/usr/lib/python2.7/dist-packages/urllib3/response.py", line 186, in read
    data = self._fp.read(amt)
  File "/usr/lib/python2.7/httplib.py", line 573, in read
    s = self.fp.read(amt)
  File "/usr/lib/python2.7/socket.py", line 380, in read
    data = self._sock.recv(left)
  File "/usr/lib/python2.7/dist-packages/urllib3/contrib/pyopenssl.py", line 188, in recv
    data = self.connection.recv(*args, **kwargs)
ZeroReturnError
```

larsblumberg commented:
I had to do a

`sudo pip install -U pip`

before installing bluepy on a fresh Raspbian. Otherwise I get

```
Downloading/unpacking https://github.com/IanHarvey/bluepy/archive/master.zip
  Downloading master.zip (unknown size): 462kB downloaded
Cleaning up...
Exception:
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/pip/basecommand.py", line 122, in main
    status = self.run(options, args)
  File "/usr/lib/python2.7/dist-packages/pip/commands/install.py", line 290, in run
    requirement_set.prepare_files(finder, force_root_egg_info=self.bundle, bundle=self.bundle)
  File "/usr/lib/python2.7/dist-packages/pip/req.py", line 1198, in prepare_files
    do_download,
  File "/usr/lib/python2.7/dist-packages/pip/req.py", line 1376, in unpack_url
    self.session,
  File "/usr/lib/python2.7/dist-packages/pip/download.py", line 572, in unpack_http_url
    download_hash = _download_url(resp, link, temp_location)
  File "/usr/lib/python2.7/dist-packages/pip/download.py", line 433, in _download_url
    for chunk in resp_read(4096):
  File "/usr/lib/python2.7/dist-packages/pip/download.py", line 421, in resp_read
    chunk_size, decode_content=False):
  File "/usr/lib/python2.7/dist-packages/urllib3/response.py", line 256, in stream
    data = self.read(amt=amt, decode_content=decode_content)
  File "/usr/lib/python2.7/dist-packages/urllib3/response.py", line 186, in read
    data = self._fp.read(amt)
  File "/usr/lib/python2.7/httplib.py", line 573, in read
    s = self.fp.read(amt)
  File "/usr/lib/python2.7/socket.py", line 380, in read
    data = self._sock.recv(left)
  File "/usr/lib/python2.7/dist-packages/urllib3/contrib/pyopenssl.py", line 188, in recv
    data = self.connection.recv(*args, **kwargs)
ZeroReturnError
```

shaftoe commented:
Hi Lars, I'll try to clarify a bit why I need this (kind of) change in the first place, context is here: https://github.com/balloob/home-assistant/pull/1487 . Basically, I need to have the Nuimo class (and dependencies) available as a Python package installable via Pip to be able to write the integration
* I've got no preference with the names at all, so I'll rename it as you suggest
* I'm not Python packaging expert but I think it really needs to live in a subfolder. I didn't dig very deeply into the Pip docs yet to be honest, so maybe it's not
* yes, the setup.py is run by `pip install` command, so that's why it's needed. You can find all the details about Pip format here: https://pip.pypa.io/en/stable/reference/pip_install/ . It's a lot and I didn't find the will to go through it properly yet, but I'm afraid it's necessary for the SDK (can't wait!) to comply to that to be able to make use of it in 3rd party projects like Home Assistant. For example the devs there asked me to provide `bluepy` as a dependency for nuimo, hence removing the bluepy dependency from the component and leaving only (for now) nuimocore as dependency, and let Pip do the rest

larsblumberg commented:
Thanks for clarifying! After you've made the changes and add this additional build step (that pip thing from my last comment), we are all set up to merge.

shaftoe commented:
Hi @larsblumberg , what's the ETA for the new SDK? I'd like to know how to proceed with https://github.com/balloob/home-assistant/pull/1487

larsblumberg commented:
Hi @shaftoe, great to see your progress on home-assistant! We have a working demo of pygattlib with Nuimo but due to limited resources out hasn't become a full featured SDK yet. If we share the demo would you be interested to continue on top of that?

shaftoe commented:
Sure! Demo is enough, Home Assistant integration is my hackathon project so I'd like to have something to show before the deadline, but I don't need that to be validated by Home Assistant upstream before the deadline, so please proceed

larsblumberg commented:
Alright, will share it with you soon!

shaftoe commented:
Great news! :+1:

larsblumberg commented on setup.py:
```diff
@@ -0,0 +1,12 @@
+setup(name='nuimo-bluepy',
+      version='0.0.1',
+      description='Python API bridge to a Nuimo device',
+      url='https://github.com/shaftoe/nuimo-raspberrypi-demo/',
```
This one should probably point to `getsenic` instead.

larsblumberg commented:
Dear Alexander, thanks for this PR. I've just tested it and it works great. Can you just change the line mentioned above before merging it?

larsblumberg commented:
For the pygattlib I don't have good news yet. I had a misunderstanding with the developer who worked on a asynchronous nuimo library, e.g. not having to call blocking methods such as `waitForNotifications`. He instead implemented an asynchronous poller that would wait for nuimo events.

How do you handle nuimo events in your project? Do you use `waitForNotifications`? How do you make it asynchronous in your case?

shaftoe commented:
Yes, I made use of `waitForNotifications` in a background thread (using threading Python module), and that's actually ok in my case because the Nuimo poller won't be run in the main thread anyway. You could have a look at the actual implementation here: https://github.com/shaftoe/home-assistant/blob/1b18a809df3405ff50e2aec60c315af4372d0c2c/homeassistant/components/nuimo.py#L8

larsblumberg commented:
Hey Alex, thanks for your example!

By the way, are still interested in merging this PR? I placed comment for one line, otherwise your PR is ready to merge.

shaftoe commented:
I think that with latest PR version (d59593d90de6cce78dff9348a3fe493b4e497870) I should have addressed all your suggestions, or am I missing something?
By the way, I'll really need this PR when Python v3 is fully supported, so up to you if you want to merge because it could be useful for other reasons or not

larsblumberg commented:
Oh, I didn't get a change notification. Merged now!
